presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java

```diff
@@ -61,13 +62,17 @@
 
     private final SqlStageExecution stage;
     private final List<InternalNode> nodes;
-    private final List<SourceScheduler> sourceSchedulers;
+
+    // use a CopyOnWriteArrayList to prevent ConcurrentModificationExceptions
```

Comment:
Hmm... could we do something like just not call `.clear()` on the list instead? (Why do we clear() it?)

Reply:
If the intent is to short circuit the loop where we break if `cancelled`, isn't that the point of the cancelled flag, and it won't work because the iterator is merrily iterating over the old, un-cleared version of the list.